core/sequencer/runner/native.go

```diff
@@ -118,3 +86,39 @@ func DoMapMapLeafFn(fn MapMapLeafFn, leaves []*tpb.MapLeaf) ([]*entry.IndexedVal
 	}
 	return outs, nil
 }
+
+// ReduceMutationFn takes all the mutations for an index and an auxiliary input
+// of existing mapleaf(s) returns a new map leaf.  ReduceMutationFn must be
+// idempotent, commutative, and associative. i.e. must produce the same output
+// regardless of input order or grouping, and it must be safe to run multiple
+// times.
+type ReduceMutationFn func(msgs []*pb.EntryUpdate, leaves []*pb.EntryUpdate,
+	emit func(*pb.EntryUpdate), emitErr func(error))
+
+// DoReduceFn takes the set of mutations and applies them to given leaves.
+// Returns a list of key value pairs that should be written to the map.
+func DoReduceFn(reduceFn ReduceMutationFn, joined []*Joined, emitErr func(error)) []*entry.IndexedValue {
+	ret := make([]*entry.IndexedValue, 0, len(joined))
+	for _, j := range joined {
+		reduceFn(j.Values1, j.Values2,
+			func(e *pb.EntryUpdate) {
+				ret = append(ret, &entry.IndexedValue{Index: j.Index, Value: e})
+			},
+			func(err error) { emitErr(fmt.Errorf("reduceFn on index %x: %v", j.Index, err)) },
+		)
+	}
+	return ret
+}
+
+// DoMarshalIndexedValues executes Marshal on each IndexedValue
```

Comment:
It looks like some of the elements in the result could be nil. Is this worth documenting, or generally understood behaviour for Go functions?

Reply:
Generally nil values in go are not expected.  Typically they should be accompanied by an error. 
Maybe I'm missing something. What's the path to producing a nil value in this function?

Reply:
On line 117 (mapLeaf == nil xor err == nil) from my reading of Marshal. You handle err by emitting it, but then carry on to poke the nil mapLeaf into the slice. This means you could get a nil value in there, which might surprise anyone iterating over it.

Not sure what the right approach is here now we have the emitErr mechanism. You could either fully fail, or at least document that the result contain a subset of the expected values?

Reply:
Added a `continue` and documented the results might be a subset.